requests-for-comments/2022-03-08-full-saml.md

```diff
@@ -0,0 +1,53 @@
+# RFC: Full SAML support
+
+- **Created:** *2022-03-08*
+- **Owner:** @mariusandra (previously @paolodamico)
+- **GitHub issues:** [#5647](https://github.com/PostHog/posthog/issues/5647)
+
+> While commenting on this, please comment on specific lines if possible, to thread the discussion.
+
+## Overview
+
+Currently we support SAML only on self-hosted. More users have requested support for this feature on the Cloud version too. This RFC contains a technical proposal for how to implement full multitenant SAML support.
+
+On multitenant environments SAML can present a risk because these servers are asserting to the identity of a user, and we can't verify that a specific server has permission to assert the identity of a specific user (e.g. I could create a malicious SAML server that says I'm `admin@posthog.com` with no permissions over that login). To address this, we'll introduce domain verification through DNS so that users configuring SAML can also validate they have permissions over a specific domain.
+
+## Background
+
+From the original SAML issue on GitHub,
+> This is a compliance feature that a lot of large-scale organizations require for full widespread rollout. When working at a large org, having a centralized place for provisioning and managing users can make the onboarding/offboarding process smoother and help reduce security risks. There's a great [Slack thread](https://posthogusers.slack.com/archives/CT7HXDEG3/p1628859965082200) by a community member about the importance of SAML and [this study](https://www.onelogin.com/press-center/press-releases/97-percent-of-saas-vendors-backing-saml-based-single-sign-on) by OneLogin in which they find that up to 97% of SaaS support have support (or plans) for SAML. 
+## Proposal
+
+### Self-hosted
+1. Self-hosted instances remain with single-tenant SAML support (i.e. only one SAML provider for each instance), but we drop `SAML_ENFORCED` in favor of a `SSO_ENFORCEMENT` instance-level attribute that supports `saml`, `google-oauth2`, `github` & `gitlab`. We redirect automatically to the provider's login page if SSO is enforced (solves [#7535](https://github.com/PostHog/posthog/issues/7535)).
+
+### PostHog Cloud
+1. On PostHog Cloud create a separate model for `OrganizationConfiguration` where we track all SAML parameters (ACS URL, Entity ID, X509 certificate and 4 attribute names) at the organization-level.
+2. Before SAML can be enabled, admins have to verify ownership of a specific domain by adding a TXT record to their domain with a random 32-byte secret. An organization can verify multiple domains, but a domain may belong to only one organization. `OrganizationDomain` model for this.
```

Comment:
As this will be an enterprise feature, another simpler approach here would be to do what Notion does, we manually assign authorized domains to each organization (and validate manually).